### ISSUE TITLE
GPG key support

### ISSUE THREAD

**Issue by jaxxstorm** [bitnami-labs/sealed-secrets#156]
I'm wondering if it would be possible to add support for GPG keys to sealed-secrets.

If it was possible to specify a GPG private key and public key, it would also be possible to integrate with something like sops

**Comment by KlavsKlavsen** [comment]
And it would be possible to have a backup access - based on f.ex. yubikey access - to read secrets - if necessary.

**Comment by mkmik** [comment]
the current process for that is that you backup the private key (possibly encrypting them with gpg/yubikey/whatever) and then if needed you can use `kubeseal --recovery-unseal` to decrypt your sealed secrets.

**Comment by KlavsKlavsen** [comment]
@mkmik Problem with that is the private key can then be stolen :( (as I store it on my local computer).
IF we used GPG keys directly on the secrets.. (and as passwordstore - encode to be readable by multiple GPG recipients) - I could encode so "test-cluster-secrets-service" could read secrets f.ex. in a specific folder.. AND I (and those configured as recipients) can decrypt - but NEVER steal my gpg key to get access - as its stored on the yubikey.
THEY COULD only steal the secret gpg key inside the "test-cluster-secrets-service" namespace.. but if they have that access they own all secrets in that cluster anyways..
and in my passwordstore - I could easily see all secrets affected by such a compromise (as it would be all in that 'test-cluster-secrets' folder.
Did I mention I was a fan of NOT having private keys that can be stolen? several big security compromises of Mozilla and others - were due to developer secret keys (like ssh keys) being stolen :) - and since an SSH key can derived from a gpg key - I also use gpg on yubikey for SSH private key - avoiding that issue :)

So in short - I do see real security benefit to supporting a setup "like passwordstore.org" - as secrets store.
One implementation COULD be to allow a passwordstore.org as a "backend/git repo" source of secrets ?

**Comment by mkmik** [comment]
yep it's a tradeoff;

The sealed secret private key is as secure as the cluster it runs in.

1. If you *lose* the private key, you cannot decrypt your secrets.
2. If you *leak* the private key, everybody can decrypt your secrets.

In order to protect yourself from the chance of (1) you increase the risk of (2).

Secure backups are a general problem, which can be approached with encryption, but, it's turtles all way down:
where do you store the private key that encrypted your backup!?

A yubikey is a good place to store the private key that can be used to decrypt your backup.
Yes the backup contains a private key; yes somebody can read that encrypted file, no, they cannot brute force it without your gpg private key (which as you correctly point out is derived from a hardware key in your yubikey).

So, yes; encrypting the individual secrets with multiple keys would be an elegant solution (whether GPG is the technology of choice is another topic), we even have an issue for that #263.

While we don't have that, performing a secure backup (possibly on the cluster so that the cleartext never leaves the cluster) is the second best option and it can be done today.

---

(Also, keep in mind that (2) can always happen, no matter how well you encrypt your backup. Sealed secrets is designed to encrypt things like database passwords, oauth secrets and other stuff that you're supposed to rotate periodically!)

(btw, fun project a colleague of mine is playing with: https://github.com/wojciechka/openscad-designs/tree/main/password-storer)
